### ISSUE TITLE
No module named 'docplex'

### ISSUE THREAD

**Issue by arani-mohammad** [IBMDecisionOptimization/docplex-examples#12]
Hi @vlkong ,
I have been having another problem that I appreciate your help in advance if you could guide me another time.
As you may remember from our previous correspondence, 

[https://github.com/IBMDecisionOptimization/docplex-examples/issues/11]

I changed my PC for better run time, and I have done the same as I explained in the above link. Now my problem is that I can not run the same code. And the error is: 

````
ModuleNotFoundError                       Traceback (most recent call last)
<ipython-input-6-d5ae0bcc406c> in <module>
      1 import setup
----> 2 from docplex.mp.model import Model

ModuleNotFoundError: No module named 'docplex'
````

I am using the latest version of Jupyter Notebook for Windows 64 bit. 
The only difference here from the above link is that on the new machine which the latest anaconda installed the python version is 3.7.

Thank you.
M. A.

**Comment by vlkong** [comment]
Hi @arani-mohammad,

You need to install docplex. In your notebook, create a new cell containing:
```
!pip install docplex
```
Depending on your permissions on changing the Jupyter installation, you might need to change that to `!pip install --user docplex` instead.

You run that cell, then restart your kernel. This needs to be done only once, so you can remove the cell after you run it.

**Comment by arani-mohammad** [comment]
Hi @vlkong ,
Thank you very much for your time and responding to my question, I appreciate your help.
As you recommended, first I installed docplex by 
`!pip install docplex`
then I came across the limitation of the community version
`DOcplexLimitsExceeded: **** Promotional version. Problem size limits exceeded, CPLEX code=1016`
then I uninstalled the community version by
`!pip uninstall docplex`
and then I added the following block of code as I had done before and I had solved my OR problem
`import setup
from docplex.mp.model import Model`
[https://github.com/IBMDecisionOptimization/docplex-examples/issues/11](url)
But now I am getting another error that I cannot go around it even by re-installing the community version as before
`
TypeError: can't pickle dict_keys objects
`

`
---------------------------------------------------------------------------
TypeError                                 Traceback (most recent call last)
<ipython-input-102-2809b40e2c11> in <module>
      1 mdl.parameters.timelimit=60
----> 2 solution=mdl.solve(log_output=True)
      3 print("*****************************************************************")
      4 # print(solution)
      5 print("*****************************************************************")

C:\ProgramData\Anaconda3\lib\site-packages\docplex\mp\model.py in solve(self, **kwargs)
   3614         lex_timelimits = kwargs.pop('lex_timelimits', None)
   3615         lex_mipgaps = kwargs.pop('lex_mipgaps', None)
-> 3616         context = self.prepare_actual_context(**kwargs)
   3617 
   3618         if lex_timelimits is not None and len(lex_timelimits) == 1:

C:\ProgramData\Anaconda3\lib\site-packages\docplex\mp\model.py in prepare_actual_context(self, **kwargs)
   3569             elif argname != "context" and argval is not None:
   3570                 if not cloned:
-> 3571                     context = context.clone()
   3572                     cloned = True
   3573                 context.update_key_value(argname, argval)

C:\ProgramData\Anaconda3\lib\site-packages\docplex\mp\context.py in clone(self)
    419         # Returns:
    420         #   A deep copy of the context.
--> 421         return deepcopy(self)
    422 
    423     def update_from_list(self, values, logger=None):

C:\ProgramData\Anaconda3\lib\copy.py in deepcopy(x, memo, _nil)
    178                     y = x
    179                 else:
--> 180                     y = _reconstruct(x, memo, *rv)
    181 
    182     # If is its own copy, don't memoize.

C:\ProgramData\Anaconda3\lib\copy.py in _reconstruct(x, memo, func, args, state, listiter, dictiter, deepcopy)
    304             for key, value in dictiter:
    305                 key = deepcopy(key, memo)
--> 306                 value = deepcopy(value, memo)
    307                 y[key] = value
    308         else:

C:\ProgramData\Anaconda3\lib\copy.py in deepcopy(x, memo, _nil)
    178                     y = x
    179                 else:
--> 180                     y = _reconstruct(x, memo, *rv)
    181 
    182     # If is its own copy, don't memoize.

C:\ProgramData\Anaconda3\lib\copy.py in _reconstruct(x, memo, func, args, state, listiter, dictiter, deepcopy)
    278     if state is not None:
    279         if deep:
--> 280             state = deepcopy(state, memo)
    281         if hasattr(y, '__setstate__'):
    282             y.__setstate__(state)

C:\ProgramData\Anaconda3\lib\copy.py in deepcopy(x, memo, _nil)
    148     copier = _deepcopy_dispatch.get(cls)
    149     if copier:
--> 150         y = copier(x, memo)
    151     else:
    152         try:

C:\ProgramData\Anaconda3\lib\copy.py in _deepcopy_dict(x, memo, deepcopy)
    238     memo[id(x)] = y
    239     for key, value in x.items():
--> 240         y[deepcopy(key, memo)] = deepcopy(value, memo)
    241     return y
    242 d[dict] = _deepcopy_dict

C:\ProgramData\Anaconda3\lib\copy.py in deepcopy(x, memo, _nil)
    148     copier = _deepcopy_dispatch.get(cls)
    149     if copier:
--> 150         y = copier(x, memo)
    151     else:
    152         try:

C:\ProgramData\Anaconda3\lib\copy.py in _deepcopy_list(x, memo, deepcopy)
    213     append = y.append
    214     for a in x:
--> 215         append(deepcopy(a, memo))
    216     return y
    217 d[list] = _deepcopy_list

C:\ProgramData\Anaconda3\lib\copy.py in deepcopy(x, memo, _nil)
    178                     y = x
    179                 else:
--> 180                     y = _reconstruct(x, memo, *rv)
    181 
    182     # If is its own copy, don't memoize.

C:\ProgramData\Anaconda3\lib\copy.py in _reconstruct(x, memo, func, args, state, listiter, dictiter, deepcopy)
    278     if state is not None:
    279         if deep:
--> 280             state = deepcopy(state, memo)
    281         if hasattr(y, '__setstate__'):
    282             y.__setstate__(state)

C:\ProgramData\Anaconda3\lib\copy.py in deepcopy(x, memo, _nil)
    148     copier = _deepcopy_dispatch.get(cls)
    149     if copier:
--> 150         y = copier(x, memo)
    151     else:
    152         try:

C:\ProgramData\Anaconda3\lib\copy.py in _deepcopy_dict(x, memo, deepcopy)
    238     memo[id(x)] = y
    239     for key, value in x.items():
--> 240         y[deepcopy(key, memo)] = deepcopy(value, memo)
    241     return y
    242 d[dict] = _deepcopy_dict

C:\ProgramData\Anaconda3\lib\copy.py in deepcopy(x, memo, _nil)
    148     copier = _deepcopy_dispatch.get(cls)
    149     if copier:
--> 150         y = copier(x, memo)
    151     else:
    152         try:

C:\ProgramData\Anaconda3\lib\copy.py in _deepcopy_list(x, memo, deepcopy)
    213     append = y.append
    214     for a in x:
--> 215         append(deepcopy(a, memo))
    216     return y
    217 d[list] = _deepcopy_list

C:\ProgramData\Anaconda3\lib\copy.py in deepcopy(x, memo, _nil)
    178                     y = x
    179                 else:
--> 180                     y = _reconstruct(x, memo, *rv)
    181 
    182     # If is its own copy, don't memoize.

C:\ProgramData\Anaconda3\lib\copy.py in _reconstruct(x, memo, func, args, state, listiter, dictiter, deepcopy)
    278     if state is not None:
    279         if deep:
--> 280             state = deepcopy(state, memo)
    281         if hasattr(y, '__setstate__'):
    282             y.__setstate__(state)

C:\ProgramData\Anaconda3\lib\copy.py in deepcopy(x, memo, _nil)
    148     copier = _deepcopy_dispatch.get(cls)
    149     if copier:
--> 150         y = copier(x, memo)
    151     else:
    152         try:

C:\ProgramData\Anaconda3\lib\copy.py in _deepcopy_dict(x, memo, deepcopy)
    238     memo[id(x)] = y
    239     for key, value in x.items():
--> 240         y[deepcopy(key, memo)] = deepcopy(value, memo)
    241     return y
    242 d[dict] = _deepcopy_dict

C:\ProgramData\Anaconda3\lib\copy.py in deepcopy(x, memo, _nil)
    148     copier = _deepcopy_dispatch.get(cls)
    149     if copier:
--> 150         y = copier(x, memo)
    151     else:
    152         try:

C:\ProgramData\Anaconda3\lib\copy.py in _deepcopy_list(x, memo, deepcopy)
    213     append = y.append
    214     for a in x:
--> 215         append(deepcopy(a, memo))
    216     return y
    217 d[list] = _deepcopy_list

C:\ProgramData\Anaconda3\lib\copy.py in deepcopy(x, memo, _nil)
    167                     reductor = getattr(x, "__reduce_ex__", None)
    168                     if reductor:
--> 169                         rv = reductor(4)
    170                     else:
    171                         reductor = getattr(x, "__reduce__", None)

TypeError: can't pickle dict_keys objects
`
I have looked it up on the internet and some people recommended to manipulate the Jupyter Config., again I could not figure out 
[https://datalab.noao.edu/help/index.php?qa=254&qa_1=notebookapp-iopub_data_rate_limit](url)
I was wondering if you might have any suggestion for me.

**Comment by vlkong** [comment]
You can try installing cloudpickle in this case.

**Comment by arani-mohammad** [comment]
Thanks,
But I do not have a clue to how to use cloudpickle. I have been reading different documents, but could not implement it.

**Comment by vlkong** [comment]
I think this is related to https://github.com/IBMDecisionOptimization/docplex-examples/issues/14
Please reopen issue if this is not the case. Thanks !

**Comment by lanvi9110** [comment]
People ask me:
When I code the VRP model and unpack it, I get an error like below. How can I fix this error?

DOcplexException                          Traceback (most recent call last)
<ipython-input-29-428060eecfaf> in <module>
      1 mdl.parameters.timelimit=120
----> 2 sol= mdl.solve()
      3 sol.display()

~\anaconda3\lib\site-packages\docplex\mp\model.py in solve(self, **kwargs)
   4734                 return self._solve_local(context, used_clean_before_solve, lex_timelimits, lex_mipgaps)
   4735             else:
-> 4736                 return self.fatal("Cannot solve model: no CPLEX runtime found.")
   4737 
   4738     def _connect_progress_listeners(self):

~\anaconda3\lib\site-packages\docplex\mp\model.py in fatal(self, msg, *args)
   1072 
   1073     def fatal(self, msg, *args):
-> 1074         self._error_handler.fatal(msg, args)
   1075 
   1076     def fatal_ce_limits(self, *args):

~\anaconda3\lib\site-packages\docplex\mp\error_handler.py in fatal(self, msg, args)
    209         resolved_message = resolve_pattern(msg, args)
    210         docplex_error_stop_here()
--> 211         raise DOcplexException(resolved_message)
    212 
    213     def fatal_limits_exceeded(self):

DOcplexException: Cannot solve model: no CPLEX runtime found.

**Comment by vlkong** [comment]
@lanvi9110 Can you please open a new issue ?
